### DISCUSSION TITLE
Screen corruption while using a single DMA sprite on Pico Pi with ILI9341

### DISCUSSION THREAD

**Discussion by aord** [Bodmer/TFT_eSPI#1896] [General]
Many thanks Bodmer and any helpers for the fantastic library.

**I'm using a single DMA sprite to draw to the screen, but have corruption/artifacts.**
Not sure if my code is faulty? Or it is the limitations of screen? Or a fault somewhere else? 
Anybody have any ideas why?

![TFT corruption2](https://user-images.githubusercontent.com/15178107/177060763-c65a74d9-a209-47b2-83f7-428eea5775d8.png)

Screen works perfect using two DMA sprites.
Thought maybe the sprite was too big (240x320) so tested with 160x160 sprite, but still corruption in 160x160 DMA sprite area.
Strangely I noticed if I added delay (~60MS or bigger, FPS=16.23) the artifacts would go away.

My setup:
Pico Pi
320x240 ILI9341 2.4" (non touch)
Setup60_RP2040_ILI9341.h (wired SPI, same pins as described in header, same result with TFT_CS to ground and to pin20)
SPI_FREQUENCY   40000000 (40MHz, tested with 80MHz no change)

Code:

`#include <TFT_eSPI.h> 
#include <SPI.h>
TFT_eSPI tft = TFT_eSPI();  
uint16_t* sprPtr;
TFT_eSprite spr = TFT_eSprite(&tft);

void setup() 
{
    sprPtr = (uint16_t*)spr.createSprite(240, 320); 
    tft.init(); 
    tft.initDMA(); 
    tft.startWrite(); 
}

void loop() 
{
    spr.fillSprite(TFT_BLACK);
    spr.fillRect(0,0,80,320, TFT_WHITE);
    spr.fillCircle(40, 40, 38, TFT_RED);
    tft.pushImageDMA(0, 0, 240, 320, sprPtr);
    //NOTE adding delay(60); //This gives 16.23 FPS and screen now WORKING ??!?!??
}`

**Comment by Bodmer** [comment]
The problem is in the sketch, the sprite is being redrawn by the processor while it is still being rendered by the DMA engine. Either wait for the sprite rendering to complete, ie use dmaWait(), before changing it or use double buffering with 2 sprites.

**Reply by aord** [reply]
Excellent. Many thanks, added tft.dmaWait() to start of loop and now perfect.

Notes: 
Two DMA sprites is slightly faster for the loop code I used.
Running the code single DMA sprite I get 17.3 FPS, Running two DMA sprites I get 17.85. So two DMA sprites is slightly faster.
Running single DMA drawing 100 filled circles gives 8.16 FPS and two DMA 11.82 FPS.
